### PR TITLE
API - rswag integration

### PR DESCRIPTION
**Pull request by Matt-Yorkley**
#### What? Why?

Lays out an example of what a V1 API controller might look like, using the various bits we've discussed around Swagger and Serialization changes. 

There's a lot of decisions and choices inherent in this example which would affect the whole course of the API (if we're building everything else in roughly this way), so it would be good to discuss any pros and cons of the approach and decide if is this more or less how we want to do things, or if we want to make some different choices, or go in a different direction.

There's a lot to unpack here... 

### Swagger

Implements automatic Swagger documentation using on the `rswag-spec` gem. The documentation is both automatically generated based on the request spec, and automatically published, available on each instance. The API docs UI is also interactive, so you can experiment with requests directly in the browser based on the current logged in user.

There's some non-standard syntax for the parts of the spec which generates the documentation. It looks like this:
```rb
path "/api/v1/customers" do
  get "List customers" do
    tags "Customers"
    parameter name: :enterprise_id, in: :query, type: :string
    produces "application/json"
 
    response "200", "Customers list" do
      param(:enterprise_id) { enterprise1.id }
      schema CustomerSchema.collection(require_all: true)

      run_test!
    end
  end
end
```
There are some specific methods here like `tags` and `run_test!` which are part of the rswag DSL. Docs here: https://github.com/rswag/rswag

These specs create the documentation and also test the responses against JSON schemas to ensure they are correct. They don't really test behaviour or complex tests based on contexts like permissions; those kinds of tests can be added alongside them (and with more familiar/standard Rspec syntax).

### JSON Schemas

Swagger makes heavy use of JSON schemas for validating deeply-nested JSON structures. It's a well established standard with docs here: https://json-schema.org/

An example of a JSON schema that describes an array of strings:
```ruby
{
  type: :array,
  items: {
    type: :string
  }
}
```
Deeply-nested JSON schemas can get a bit unwieldy, and the `rswag` gems don't provide helpers for working with `JSON:API` standard by default, so I've added some simple structures which allow building valid schemas easily (see `JsonApiSchema` and `CustomerSchema`).  

### jsonapi-serializer

I've added the `jsonapi-serializer` gem here, which is significantly faster than `ActiveModel::Serializers` and was originally developed by Netflix. Docs here: https://github.com/jsonapi-serializer/jsonapi-serializer

### JSON:API

The `jsonapi-serializer` gem follows the `JSON:API` standard, which is documented here: https://jsonapi.org and has some nice examples here: https://jsonapi.org/examples

It's a nice standard and ensures consistency and a widely recognised format for JSON responses :+1:

### Generating docs

If you make any changes to the request specs or to the configs in the swagger helper, you can run `rails rswag`, and it'll regenerate and update the docs. These are then published under the URL `/api-docs`.

The published API  docs look like this (you can give it a try in dev or staging):

API-peek

#### What should we test?

.

#### Release notes

Changelog Category: Technical changes

### CODE REVIEW

**Comment by Matt-Yorkley** [comment]
Hi! :wave::smile:

> fast_jsonapi... not maintained anymore

Yeah, [jsonapi-serializer](https://github.com/jsonapi-serializer/jsonapi-serializer) is the active community-maintained fork of the original. It seems pretty good, and there doesn't seem to be many other contenders. We can do an update in the discourse thread for sure :+1:

> It's awesome to see that rswag-spec can actually work!!

Yeah, it's nice. The only thing that made me think twice was the extent we have to generate these json schemas in various places for rswag. I think a lot of the time they'll look almost identical (like 95% re-usable boilerplate, and 5% unique parts in the structure). I tried to deal with that nicely in [JsonApiSchema](https://github.com/openfoodfoundation/openfoodnetwork/blob/39bb53fa26b4a65758be2f44917f1ebed80704bf/app/json_schemas/json_api_schema.rb) and [CustomerSchema](https://github.com/openfoodfoundation/openfoodnetwork/blob/39bb53fa26b4a65758be2f44917f1ebed80704bf/app/json_schemas/customer_schema.rb) to avoid having totally wild amounts of duplication. It's a tradeoff, but the benefits that come with this overall rswag approach seem worth it (automatically generating the API documentation from our specs is great).

> I am not too sure about the versioning strategy, when you create v2, will you start v2 by copy pasting all things in v1 folders to a new v2 folders and start changing to create v2?

So a few months ago we moved all the pre-existing API endpoints under the `/api/v0` namespace [here](https://github.com/openfoodfoundation/openfoodnetwork/pull/6960), which came out of discussion in this little experiment: https://github.com/openfoodfoundation/openfoodnetwork/issues/6886. We also considered naming the old routes `/api/legacy` or `/api/internal`.

So all the old endpoints are considered internal / deprecated, and we'll make the new ones under `v1`, which will be the public-facing API.

> copy pasting all things

No! The exact opposite! :sweat_smile: 

Since #6960 we're now free to _not_ copy-paste all the things. So for example, we can also change the serializers and switch to the JSON:API specification for the new public API, but old bits of Angular code that are still using some of those `v0` endpoints can stay as they are (so we won't need to rewrite everything to fit with the new format). 

We can effectively start with a blank slate and very little in terms of constraints imposed by things like... weird legacy Spree code. Take this `v0` endpoint in the `TaxonsController` for example: https://github.com/openfoodfoundation/openfoodnetwork/blob/c3076f031614b3bc541db80bc73f8383e3a9c80e/app/controllers/api/v0/taxons_controller.rb#L21-L24

The only reason it's there is to service a really obscure bit of frontend javascript ([jstree](https://www.jstree.com/)) which is only included on one page in the app, and that page is only accessible by superadmin users. It's not a RESTful route, and it's only really useful in the context of that one weird (wildly outdated) jQuery plugin that dates back to Spree 0.7 (2010!). And that page doesn't even have good UX! That endpoint obviously doesn't belong in the public-facing API, and now we're free to just ignore it in `v1`.

So we can approach the new API controllers purely from the point of view of **designing a nice, clean, sensible, RESTful API that's as widely and generically useful to as broad a range of external clients as possible**, without having to worry about how to make everything fit in with weird bits of Spree or existing Angular :partying_face:

File: app/controllers/api/v1/base_controller.rb
Comment (mkllnk) [inline app/controllers/api/v1/base_controller.rb:37]: Can we remove assignments from conditionals? I would find this a lot easier to read:
```suggestion
      def authenticate_user
        @current_api_user = warden_user || api_key_user || guest_user
        invalid_api_key if @current_api_user.nil?
      end
      
      def warden_user
        request.env['warden'].user
      end

      def api_key_user
        Spree::User.find_by(spree_api_key: api_key.to_s)
      end
      
      def guest_user
         Spree::User.new if api_key.blank?
      end
```

File: app/controllers/api/v1/base_controller.rb
Comment (mkllnk) [inline app/controllers/api/v1/base_controller.rb:15]:
```suggestion
      attr_accessor :api_user
```
:question:

**Review by mkllnk — COMMENTED** [review]
I saw this pull request back in October and was pretty excited. I'm not sure about the process here. Are you just waiting for criticism and then we can discuss or how do we make decisions here? The code doesn't exactly list discussion points with alternatives. But I'll try to start that here.

* Swagger and specs: yay! :tada: The initial Swagger documentation was community driven and we had several community requests since. Swagger seems to be a popular tool. We need specs and we need documentation and I'm not aware of another tool that provides it all as well. This choice seems to be a no brainer to me unless somebody finds something better.
* JSON schema: sounds good, what exactly is the benefit? Do we need it for the Swagger integration? Will it make JSON-LD extension easier?
* jsonapi-serializer: sure thing. I would hope that Rails will catch up and add this or become as fast. But we can use this gem in the meantime.
* JSON:API: it's a good standard. The only real alternative is JSON-LD to connect us directly to the semantic web including the DFC standards. But from what I've read and saw so far, JSON:API is much closer to what we got at the moment, will be much easier to learn and faster to develop which should enable lots more community contributions. And it's not a dead end, we can still convert it to JSON-LD, maybe just with a [context](https://www.cloudbees.com/blog/json-ld-building-meaningful-data-apis). Anyway, I think that JSON:API is a good step forward and that DFC mapping should follow afterwards and it's still a possibility for v2 if not a simple extension.

I don't see anything contentious in here. Okay, pagination has been a discussion point on Slack. We can always publish without pagination first and add it on later. Extending without breaking backwards-compatibility should be allowed.

We should get this PR ready for review as soon as possible. I see this as S3. It will unblock a lot including an Australian intern who could start working on more endpoints...

File: app/controllers/api/v1/base_controller.rb
Comment (Matt-Yorkley) [reply app/controllers/api/v1/base_controller.rb:37]: Ah, nice. I pulled this method straight from our existing API base controller and didn't re-assess it. That looks much better! :heart:

Comment (luisnramos) [reply app/controllers/api/v1/base_controller.rb:37]: nice refactor @mkllnk 👏 
In the new version, is authenticate_user returning @current_api_user if @current_api_user is not nil? I'd write a unit test to make it clear.
Or maybe the last line could be written as a ternary op to make it explicit?
`@current_api_user.present? ? @current_api_user : invalid_api_key`

Comment (luisramos0) [reply app/controllers/api/v1/base_controller.rb:37]: nice refactor @mkllnk 👏 
In the new version, is authenticate_user returning @current_api_user if @current_api_user is not nil? I'd write a unit test to make it clear.
Or maybe the last line could be written as a ternary op to make it explicit?
`@current_api_user.present? ? @current_api_user : invalid_api_key`

File: app/controllers/api/v1/base_controller.rb
Comment (luisramos0) [inline app/controllers/api/v1/base_controller.rb:56]: I understand you may not want to clean up this now but moving translations out of the spree namespace would be a nice clean up 👍

File: app/controllers/api/v1/base_controller.rb
Comment (luisramos0) [inline app/controllers/api/v1/base_controller.rb:76]: is this something we actually use or is useful?

File: app/controllers/concerns/json_api_pagination.rb
Comment (luisramos0) [inline app/controllers/concerns/json_api_pagination.rb:4]: I always find cconcerns concerning. I'd move this to a service with a clear API. This way there's just a random use of the "parent" properties (pagy and params) with no clear boundaries. In this case, the ControllerPaginationService would receive pagy itself, params[:per_page], and params[:page].

File: config/initializers/rswag-ui.rb
Comment (luisramos0) [inline config/initializers/rswag-ui.rb:1]: is it a typo that one is rswag_api with underscore and this one is with a dash?

File: config/initializers/rswag_api.rb
Comment (luisramos0) [inline config/initializers/rswag_api.rb:1]: I'd not allow a single letter variable in the code base. It's config here.

File: config/initializers/rswag-ui.rb
Comment (luisramos0) [inline config/initializers/rswag-ui.rb:1]: is it a typo that one is rswag_api with underscore and this one is with a dash?

File: app/controllers/concerns/json_api_pagination.rb
Comment (luisramos0) [inline app/controllers/concerns/json_api_pagination.rb:4]: I always find cconcerns concerning. I'd move this to a service with a clear API. This way there's just a random use of the "parent" properties (pagy and params) with no clear boundaries. In this case, the ControllerPaginationService would receive pagy itself, params[:per_page], and params[:page].

File: app/controllers/api/v1/base_controller.rb
Comment (luisramos0) [inline app/controllers/api/v1/base_controller.rb:76]: is this something we actually use or is useful?

File: app/controllers/api/v1/base_controller.rb
Comment (luisramos0) [inline app/controllers/api/v1/base_controller.rb:56]: I understand you may not want to clean up this now but moving translations out of the spree namespace would be a nice clean up 👍

File: config/routes/api.rb
Comment (luisramos0) [inline config/routes/api.rb:6]: Are the docs not supposed to be in prod as well?

File: config/routes/api.rb
Comment (luisramos0) [inline config/routes/api.rb:87]: Are you disabling v1 in prod here? I see that above as well. Can you extract to a method or feature toggle?
```
def api_v1_enabled?
   !Rails.env.production?
end
```

**Review by luisramos0 — APPROVED** [review]
I reviewed this carefully, I think this could be good to go already :-)

Matt, can you please not change commit history so I can later review only the commits you make from today? Thanks!

**Comment by luisramos0** [comment]
Have we stopped using that code scan tool that run rubocop?

**Comment by mkllnk** [comment]
> Have we stopped using that code scan tool that run rubocop?

Yes, unfortunately. I raised that problem in the core-devs channel a month ago and didn't get any reply. Maybe I'll just go ahead and add it again.

**Comment by jibees** [comment]
> > Have we stopped using that code scan tool that run rubocop?
> 
> Yes, unfortunately. I raised that problem in the core-devs channel a month ago and didn't get any reply. Maybe I'll just go ahead and add it again.

I have absolutely no opinion on the process of reactivating rubucop, and whether it is worth reactivating., since I can't remember why it has been deactivated.

**Comment by luisramos0** [comment]
>> whether it is worth reactivating

The tool is fundamental in giving visibility during code review over rubocop issues. Without it, rubocop issues will be ignored. Rubocop is fundamental in keeping the code base clean.
The goal is really to give visibility over rubocop issues during code review.

File: config/initializers/rswag_api.rb
Comment (Matt-Yorkley) [reply config/initializers/rswag_api.rb:1]: Agreed! I think this was generated automatically from a setup rake task provided by the gem.

File: config/initializers/rswag-ui.rb
Comment (Matt-Yorkley) [reply config/initializers/rswag-ui.rb:1]: Pretty sure this was generated automatically by a rake task as well. We can name them whatever we like though. I guess the underscore is a bit more consistent with our general naming schemes. I'll update it :+1:

**Comment by jibees** [comment]
> Rubocop is fundamental in keeping the code base clean.
So it should be a no-brainer ;)

**Comment by Matt-Yorkley** [comment]
Thanks @mkllnk :heart:

> Are you just waiting for criticism and then we can discuss or how do we make decisions here? 

Yeah, the idea was to kickstart the discussion. It feels a bit like 4 or 5 different proposals all at once here, it's a lot! I wasn't 100% sure of my _own_ opinion on all these bits until I could see a concrete example of them all combined together.

So yeah, if there's anything here that seems like a bad idea or if anyone has some alternate suggestions for how to do things, now's a good time to raise objections or alternative proposals :+1:

Otherwise if we more or less have a consensus on taking this approach, then that's great and we've got some clarity on the details of how we're going to implement the API (which is where I was trying to get us to). :slightly_smiling_face: 

Some of these details would potentially be tricky to change later on. Like... if we got half way through and then changed our minds it might be difficult (or expensive) to suddenly change direction. I'm just trying to make sure we have a really clear **blueprint** before we start building things.

>  JSON schema: sounds good, what exactly is the benefit? Do we need it for the Swagger integration? 

Yeah, these `rswag` gems for the Swagger integration make use of JSON schemas for a couple of key things; validating the responses in specs (eg; if some part of the object is supposed to be an array of strings, but the response has an array of integers, it's not correct) but also for generating the documentation (telling the user what they can expect the objects in the responses to look like in terms of structure and types, etc).

> jsonapi-serializer

One thing worth noting here is that using the `jsonapi-serializer` gem and following the `JSON:API` standard go hand in hand. It's not just a _generic_ JSON serializer, it generates all of it's output very explicitly in line with the `JSON:API` format.

> We can always publish without pagination first and add it on later. Extending without breaking backwards-compatibility should be allowed.

I'm not sure what you mean? The pagination is included here already on the index action.

File: app/controllers/api/v1/base_controller.rb
Comment (Matt-Yorkley) [reply app/controllers/api/v1/base_controller.rb:76]: I threw this in while I was doing a little bit of tidying up (whilst mostly just copying this code across from the old BaseController). 

It's really useful both in development and when running unit tests (while doing TDD for example). Normally you just get an error message with no context, and then you have to spend a bunch of extra time digging for more information. If you get it straight away it actually makes it a lot faster to fix things. It doesn't interfere with the content in the top-level `errors` key.

```
{
  errors: <error message>,
  meta: {
    backtrace: <backtrace lines> # Only in dev/test
  }
}
```

Comment (luisnramos) [reply app/controllers/api/v1/base_controller.rb:76]: ah, nice! I remember having that problem.

File: config/routes/api.rb
Comment (Matt-Yorkley) [reply config/routes/api.rb:87]: When I was putting this together I wasn't thinking we would immediately merge it and make these endpoints live. We might want to work on some more API controllers before doing more of a public "launch" for example, or we might spot some small changes we'd like to make whilst working on the _second_ or _third_ `v1` API controller, in which case we could adjust things a little bit without any problems and make them live at a slightly later date. 

That was roughly my thinking here :man_shrugging:

**Comment by mkllnk** [comment]
> JSON schema

Cool. Sounds like a very good idea.

> publish without pagination first and add it on later

That was a stupid thought of mine. I thought that we could remove the pagination from this PR because it's still contentious. And I thought that additions would be okay within the same API version. But I forgot that introducing pagination would change the default output when no pagination page parameter is given. I guess that we would then show only the first page and that's a breaking change. So we have to decide about that now.

I'm happy with your current approach to pagination. It's very established. If we want to adopt a different kind of pagination, we *can* add that on, I think (while keeping the old pagination). 50 records per page seems a bit low at first but then it's very efficient and we can always query more if we want.

Actually, I don't like how the concern is looking at that constant. It looks like a controller is supposed to override that constant. Maybe it would be better to remove the pagination from the BaseController and include it explicitly, initialising defaults like that. For example: `include Pagination, per_page_default: 50`

**Comment by Matt-Yorkley** [comment]
@mkllnk 

> If we want to adopt a different kind of pagination, we can add that on

Ah, like a pointer-based pagination system? Yeah, we could explore implementing that option. It would be more (technically) accurate in some cases, but I find them slightly more confusing from an end-user perspective, because you just end up with feedback like `page: 2, pointer: uyd9828y39y9` which isn't self-explanatory unless you already have a strong understanding of how pointer-based pagination works.

> 50 records per page seems a bit low at first but then it's very efficient and we can always query more if we want

Yeah we had a brief chat about this in Portugal. The client will be able to request a _different_ number of results per page if they want it, but keeping a relatively low number for clients that don't specify anything is probably a good idea, right? We considered bumping the default up to 100, but actually 50 seems fairly reasonable (and will keep the load on the server to a minimum for random requests).

> I don't like how the concern is looking at that constant. It looks like a controller is supposed to override that constant. 

Yeah I threw that in without thinking about it too much, with the idea that we might set different defaults for different controllers. Thinking about it now, it might be cleaner to just have one consistent default across all controllers (and not bother putting any code in place to allow individual controllers to override the default). What do you think?

That reminds me; we also discussed setting a hard _maximum_ `per_page`, for cases where the user supplies a custom value in the request. Like 500 max? 1000 max?

> `include Pagination, per_page_default: 50`

Hmmm... I'm not sure that syntax works for module includes in classes. There are some close [alternatives](https://medium.com/@eric.programmer/arguments-for-included-modules-in-ruby-8056b9fa2743) but they look a bit funky, eg:
```rb
include Pagination.new(per_page_default: 50)
```

Anyway, shall we stick with one default that's not customised? And shall we set a maximum as well?

**Comment by lin-d-hop** [comment]
The decision we came to on the hard max was 200, [as noted here.](https://community.openfoodnetwork.org/t/api-api-v1-discussions-from-the-portugal-delivery-team-mini-gathering/2456)

If we want to change that please update the forum post :pray:

File: app/controllers/concerns/json_api_pagination.rb
Comment (mkllnk) [inline app/controllers/concerns/json_api_pagination.rb:50]: Great! A method like this will always allow us to implement something different here or override the method in a controller without messing with constants! :+1:

**Review by mkllnk — COMMENTED** [review]
I'm really happy with this.

Just some rubocop issues.

**Comment by Matt-Yorkley** [comment]
I added a couple of small commits to tidy some bits up :+1:

I think we can merge this for now (so we don't leave the branch sitting around for too long) but we should keep this new endpoint disabled in production as it needs a couple of other bits before we can start using it, eg:
- there's some changes proposed to the `customers` database table which are relevent
- there's some extra attributes and details we need to add in the `CustomerSerializer` for it to fully meet the acceptance criteria

Other than that I think we've got a solid base to start from here.

**Review by mkllnk — COMMENTED** [review]
Great! I would love to get our code style up to date so that we can rely on the linter on all new PRs. Can you either fix more issues or update the todo list?

**Comment by Matt-Yorkley** [comment]
It looks like the current Rubocop linter ignores our `rubocop_todo.yml`...

**Comment by mkllnk** [comment]
> It looks like the current Rubocop linter ignores our rubocop_todo.yml...

What? Useless. Any config option for that?

Hm, it's running only on added code which probably means that it can't reference file names at the same time. Maybe `filter_mode: file` would help. Otherwise we can switch to `nofilter`. That would mean that it checks all violations, not just additions from that PR but if we keep the todo list up-to-date then that shouldn't be a problem.

**Comment by mkllnk** [comment]
Hi Matt! I think that we decided in the delivery circle that I continue this task. I will copy this branch and start working on that if that's okay with you. Just mentioning to avoid parallel work on the same thing.

**Comment by Matt-Yorkley** [comment]
> I will copy this branch and start working on that if that's okay with you.

Sure :ok_hand:

TBH I think we could just merge what's here, as it's explicitly toggled off in production. It was paused to wait for the Customers table data migration, but maybe it's not strictly blocked by that? Those changes could happen in parallel rather than in sequence. :man_shrugging: 

What do you think?

**Comment by mkllnk** [comment]
> TBH I think we could just merge what's here, as it's explicitly toggled off in production. It was paused to wait for the Customers table data migration, but maybe it's not strictly blocked by that? Those changes could happen in parallel rather than in sequence. 

I was wondering that, too. But I found that I needed to get more familiar with the code and how the different parts interact. So taking your pull request apart was the best way for me to fully understand what's happening. It also makes the chunks easier to review for others and I hope that the history will show the intent of different code parts better. I know, a lot of Git history is never looked at but do sometimes search for the origins of code parts and it really helps my understanding.

**Comment by mkllnk** [comment]
I created a rebased version in #8891 but even without the name splitting it's not ready yet. I just tested permissions and without being logged in and without giving an enterprise id I just get all customer records... :boom:

**Comment by mkllnk** [comment]
Another question: What was the purpose of the production toggle?

I would imagine that once the name splitting is done and we have the permissions sorted that we can activate this in production, right? We will need to test with real world data.

**Comment by mkllnk** [comment]
> without giving an enterprise id I just get all customer records

Correction: As guest user I get all customer records that have `nil` as user id...

File: app/controllers/api/v1/base_controller.rb
Comment (mkllnk) [inline app/controllers/api/v1/base_controller.rb:49]: Are we sure we want to publish the message? It could expose sensitive information.

Comment (Matt-Yorkley) [reply app/controllers/api/v1/base_controller.rb:49]: It only gets shown in development environment (in the `#json_api_error` method).

Comment (mkllnk) [reply app/controllers/api/v1/base_controller.rb:49]: The backtrace was hidden in production but the message was still passed on. I added a commit to my branch:
https://github.com/openfoodfoundation/openfoodnetwork/commit/7578085ec7af951b1ff11c96fe351e98a966d9d7

**Comment by mkllnk** [comment]
Closing in favour of https://github.com/openfoodfoundation/openfoodnetwork/pull/8891.

**Pull request closed without merging**
